main.go

```diff
@@ -1,18 +1,223 @@
+	data := dbQueryAllGuitars()
+	w.Header().Set("Content-type", "application/json")
+	fmt.Println("Get single guitar endpoint hit")
+	err := json.NewEncoder(w).Encode(data)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+}
+
+// POST request INSERTING a guitar to the database
+func newGuitar(w http.ResponseWriter, r *http.Request) {
+	db := dbConnection()
+	defer db.Close()
+	w.Header().Set("Content-type", "application/json")
+
+	var guitar Guitar
+	err := json.NewDecoder(r.Body).Decode(&guitar)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	sqlStatement := `
+	INSERT INTO guitars (name, brand_id, year, description)
+	VALUES ($1, $2, $3, $4)
+	returning id`
+	id := 0
+	err = db.QueryRow(sqlStatement, guitar.Name, guitar.Brand_id, guitar.Year, guitar.Description).Scan(&id)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprintf(w, "Item with ID %d was created", id)
+}
+
+// DELETE request
+func deleteGuitar(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-type", "application/json")
+	db := dbConnection()
+
+	// To retrieve a particular record form the database, we need to pass an id paremeter to the URL. We will use the following methods and assign it to the urlId variable
+	urlId := r.URL.Query().Get("id")
+
+	// To add a layer of security, we will cast the urlId param to an integer from a string. This will be passed into the database query below.
+	urlIdInt, err := strconv.Atoi(urlId)
+	if err != nil {
+		panic(err)
+	}
+
+	sqlStatement := "DELETE FROM guitars WHERE id = $1;"
+
+	res, err := db.Exec(sqlStatement, urlIdInt)
+	if err != nil {
+		panic(err)
+	}
+	count, err := res.RowsAffected()
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(count)
+	defer db.Close()
+}
+
+// Update request
+func updateGuitar(w http.ResponseWriter, r *http.Request) {
+	db := dbConnection()
+	defer db.Close()
+
+	w.Header().Set("Content-Type", "application/json")
+
+	urlId := r.URL.Query().Get("id")
+
+	urlIdInt, err := strconv.Atoi(urlId)
+
+	if err != nil {
+		panic(err)
+	}
+
+	var guitar Guitar
+
+	err = json.NewDecoder(r.Body).Decode(&guitar)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	sqlStatement := `
+	UPDATE guitars
+	SET name = $2, brand_id = $3, year = $4, description = $5
+	WHERE ID = $1;
+	`
+	_, err = db.Exec(sqlStatement, urlIdInt, guitar.Name, guitar.Brand_id, guitar.Year, guitar.Description)
+	if err != nil {
+		panic(err)
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+}
+
+func homePage(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "GuitarAPI Project Home Page")
+	fmt.Println("Endpoint Hit: Home Page")
 }
 
 func handleRequests() {
 	http.HandleFunc("/", homePage)
+	http.HandleFunc("/guitar", getSingleGuitar)
+	http.HandleFunc("/guitars", getAllGuitars)
+	http.HandleFunc("/guitar/create", newGuitar)
```

Comment:
This is REST-ful! Good job!

Here’s a really important lesson that’s at the heart of API design, and is core to what you’ll be doing as a backend dev (and will come up in interviews a lot): considering **path matching** and **methods**.

### Path matching

Note by using `/guitar/{id}` and `/guitar/create`, you technically have a conflict (assuming same method, which we’ll get back to), because they both occupy the URL structure of `/guitar/{path}`. Even though practically-speaking you won’t have a guitar with ID `create`, it’s still important to call out the conflict nonetheless because there could be other scenarios where a conflict is more likely to happen.

**Path matching** and **methods** are your two tools to resolve conflicts in an API design.

“Path matching” may not be the right term, but in many server systems it’s determining the priority in which a path will get routed to the right service (or Go function, in  our case). Consider the following:

- Say that `http.HandleFunc("/guitar", getSingleGuitar)` gets top priority—we could be able to handle `/guitar/1`, `/guitar/2`, and `/guitar/3` just fine.
- But as soon as `/guitar/create` happened, our path matching may determine `create` to incorrectly be an ID and we throw an error (no guitar with ID `create` found)
- To resolve this, we simply give `…/create`, `…/update`, and `…/delete` first dibs on the URL, because those are **static** (they are predictable words that never change). In general, **static routes usually should take priority**. Then, for `/guitar/{anything}`, if `{anything}` is one of those three exact words, that gets handled properly; otherwise, we’ll try and call `getSingleGuitar` and see what happens.

That’s one way to solve the problem! But the other is using **methods**.

### Methods

HTTP Methods ([docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods)) are the type of request a browser is sending. They can be `GET`, `HEAD`, `POST`, `PUT`, `DELETE`, `CONNECT`, `OPTIONS`, `TRACE`, and `PATCH`. For most intents and purposes, these have been around since forever, and will be around forever. They are core to how the internet (and APIs) work.

For most API work, you can focus on half of them while ignoring the rest:

| Method   | Typically used for                    |
|:---------|:--------------------------------------|
| `GET`    | **Read** single or multiple resources |
| `PUT`    | **Create** single resource            |
| `PATCH`  | **Update** single resource            |
| `DELETE` | **Delete** single resource            |
| `POST`   | Other operations                      |

✨ This is just a _convention_, not a limitation of the specification. You are free to use the methods however you want! But just know that most developers will _expect_ behavior something like this, so the more “creative” you get, the more confusing your API will be to consume.

So another way to prevent conflict would be to throw methods into the mix, like so:

- `GET /guitar/{id}` - Get single guitar
- `GET /guitars` - Get all guitars
- `PUT /guitar` - Create single guitar (_note: typically, the server assigns the ID, not the API consumer_)
- `PATCH /guitar/{id}` - Update single guitar (note this doesn’t conflict with `GET`!)
- `DELETE /guitar/{id}` - Delete single guitar (also doesn’t conflict!)

Now we don’t have need for `…/create`, `…/update`, etc. because it’s the method that handles the operation.

Of course, in your original example, you probably would have used `POST /guitar/create` and `GET /guitar/2`, which wouldn’t have been a conflict. So there’s no “right way” to solve the problem. But a general rule of thumb is for a good API to:

- **Limit** the amount of total endpoints, while
- **Maximizing** what each individual endpoint can do

And to that end, using methods well are a good design.